### PR TITLE
Fix Issue 627

### LINKED ISSUE #627: Expr sorting failure
The new expr sorting fails with a graph like below:

```
TEST(NVFuserTest, TMP) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(3);
  fusion.addInput(tv0);
  auto tv1 = add(tv0, new Double(1));
  auto tv2 = add(tv1, new Double(1));

  auto tv3 = makeSymbolicTensor(3);
  fusion.addInput(tv3);
  auto tv4 = add(tv3, new Double(1));

  auto tv5 = add(tv2, tv4);
  fusion.addOutput(tv5);

  tv2->computeAt(tv5, 2);
  tv4->computeAt(tv5, 3);

  fusion.printMath();
  fusion.printKernel();
}
```
```
unknown file: Failure
C++ exception with description "std::all_of( groups_.begin(), groups_.end(), [](std::unique_ptr<ExprGroup>& sg) { bool b = sg->producerEdges().empty() && sg->consumerEdges().empty(); return b; }) INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp":800, please report a bug to PyTorch. Couldn't succcessfully sort out the fusion expressions. There are remaining connections of the heirarchical segmentation which should have been flattened to a single ordered group, or disjoint ordered groups.
```

The algorithm we use determines that none of the initial expr groups can be merged. `tv4` and `tv5` should be merged, but their levels are apart more than 1. No merging is possible along the path from `tv1`, `tv2` and `tv5` initially due to the CA position of each tensor.

This also happens with the `test_batch_norm` Python test, which is failing as of now.

### PR DESCRIPTION
**Pull request by csarofeen**
Fixes https://github.com/csarofeen/pytorch/issues/627

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp [hunk 1]
```diff
@@ -769,10 +828,52 @@ bool ExprSegmentationSorter::supportedMerge(ExprGroup* sg1, ExprGroup* sg2) {
     return false;
   }
 
+  if (GpuLower::current()->caLoopMap().areMapped(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: This branch does not seem to have any effect.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp [hunk 2]
```diff
@@ -118,9 +118,13 @@ class ExprGroup {
   // Returns all neighbors, producers and consumers
   std::vector<ExprGroup*> getNeighbors();
 
-  // Look at all neighbors of this and return who this could merge with based on
-  // level values of this, neighbors, and merged neighbors of neighbors
-  std::vector<ExprGroup*> getMergeCandidates();
+  // Return neighbors of this proven to be safe nodes to merge with in regards
+  // to maining an acyclic graph. This looks at, neighbors  if merged, neighbors
+  // level, and merged neighbors of neighbors level. If fallback_mode_enabled
+  // will return the inverse set of ExprGroups that are not proven to be safe
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: "that are not proven" -> "that are proven"?

**Review by naoyam — APPROVED** [review]
LGTM. Just left a couple of minor comments.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Thanks, was left over from debug printing.

**Pull request merged**